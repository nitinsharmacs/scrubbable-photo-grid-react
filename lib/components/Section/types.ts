import type { GridConfig, SectionConfig, SectionType } from 'lib/types';

export interface SectionProps {
  section: SectionType;
  config: GridConfig;
  map: SectionConfig;
}
