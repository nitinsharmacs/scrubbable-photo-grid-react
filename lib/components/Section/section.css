.section {
  background-color: rgb(233, 231, 231);
  position: absolute;
  contain: layout;
}
