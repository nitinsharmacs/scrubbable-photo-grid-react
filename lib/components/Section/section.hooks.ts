import { produce } from 'immer';
import type { SectionCtx } from 'lib/components/Section/types';
import type {
  SegmentClickEvent,
  SelectHandler,
} from 'lib/components/Segment/types';
import { useCallback, useState } from 'react';

export const toggleSelectInCtx = (
  ctx: SectionCtx,
  segmentId: string,
  tilesId: string[]
): SectionCtx => {
  const segment = ctx.segments[segmentId];
  const tiles = Object.values(segment.tiles);

  // select all mode
  if (tilesId.length === tiles.length) {
    let status = true;
    if (segment.selected) {
      status = false;
    }

    segment.selected = status;

    tilesId.forEach((tileId) => {
      segment.tiles[tileId].selected = status;
    });

    return ctx;
  }

  // single select mode

  tilesId.forEach((tileId) => {
    segment.tiles[tileId].selected = !segment.tiles[tileId].selected;
  });

  segment.selected =
    tiles.filter((tile) => tile.selected).length === tiles.length;

  return ctx;
};

export const useSectionContext = (
  initialContext: SectionCtx
): [
  SectionCtx,
  { selectHandler: SelectHandler; clearSelection: () => void }
] => {
  const [ctx, updateCtx] = useState<SectionCtx>(initialContext);

  const selectHandler = useCallback(
    ({ segmentId, tilesId }: SegmentClickEvent) => {
      updateCtx(
        produce((prevCtx) => toggleSelectInCtx(prevCtx, segmentId, tilesId))
      );
    },
    []
  );

  const clearSelection = useCallback(() => {
    updateCtx(initialContext);
  }, []);

  return [ctx, { selectHandler, clearSelection }];
};
