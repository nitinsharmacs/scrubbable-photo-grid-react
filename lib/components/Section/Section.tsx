import type { SectionProps } from 'lib/components/Section/types';

import './section.css';

const Section = ({ section, config, map }: SectionProps) => {
  return (
    <div
      className='section'
      id={section.sectionId}
      style={{
        width: `${config.containerWidth}px`,
        height: `${map.height}px`,
        top: `${map.top}px`,
      }}
    ></div>
  );
};

export default Section;
