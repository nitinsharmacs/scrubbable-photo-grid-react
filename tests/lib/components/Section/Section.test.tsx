import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import Section from 'lib/components/Section/Section';
import type { GridConfig, SectionConfig, SectionType } from 'lib/types';

describe('Section', () => {
  it('should render the section', () => {
    const gridConfig: GridConfig = {
      containerWidth: 800,
      segmentMargin: 10,
      sectionMargin: 20,
      targetRowHeight: 150,
    };

    const section: SectionType = {
      sectionId: 'sec1',
      totalImages: 1,
      segments: [],
    };

    const sectionConfig: SectionConfig = {
      top: 20,
      height: 150,
      visible: false,
      index: 0,
    };
    const { container } = render(
      <Section section={section} map={sectionConfig} config={gridConfig} />
    );

    expect(container).toMatchSnapshot();
  });
});
