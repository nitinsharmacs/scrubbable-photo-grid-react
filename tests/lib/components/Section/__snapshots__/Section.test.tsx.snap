// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Section > should render the section 1`] = `
<div>
  <div
    class="section"
    id="sec1"
    style="width: 800px; height: 150px; top: 20px;"
  />
</div>
`;
